File: app/src/main/java/com/xnote/app/domain/text/FtsIndexText.kt
package com.xnote.app.domain.text

// -- Functions

object FtsIndexText {
    fun prepare(text: String): String {
        if (text.isEmpty()) return text
        val output = StringBuilder(text.length * 2)
        var index = 0
        var previousCodePoint = -1
        var previousIsCjk = false
        while (index < text.length) {
            val codePoint = text.codePointAt(index)
            val isCjk = isCjkCodePoint(codePoint)
            val isWhitespace = Character.isWhitespace(codePoint)
            if (output.isNotEmpty() && !isWhitespace && previousCodePoint != -1 && !Character.isWhitespace(previousCodePoint)) {
                val needsSeparator = (previousIsCjk && isCjk) ||
                    (previousIsCjk && !isCjk) ||
                    (!previousIsCjk && isCjk)
                if (needsSeparator) {
                    output.append(' ')
                }
            }
            output.appendCodePoint(codePoint)
            previousCodePoint = codePoint
            previousIsCjk = isCjk
            index += Character.charCount(codePoint)
        }
        return output.toString().trim().replace(Regex("\\s+"), " ")
    }

    fun matchQuery(rawQuery: String): String? {
        val prepared = prepare(rawQuery.trim())
        if (prepared.isEmpty()) return null
        val sanitized = prepared
            .replace("\"", " ")
            .replace(Regex("[{}()*?:^]"), " ")
            .replace(Regex("\\s+"), " ")
            .trim()
        if (sanitized.isEmpty()) return null
        return "\"$sanitized\""
    }

    private fun isCjkCodePoint(codePoint: Int): Boolean {
        return codePoint in 0x3400..0x4DBF ||
            codePoint in 0x4E00..0x9FFF ||
            codePoint in 0xF900..0xFAFF ||
            codePoint in 0x20000..0x2FA1F ||
            codePoint in 0x3040..0x30FF ||
            codePoint in 0xAC00..0xD7AF
    }
}

File: flutter/lib/domain/text/fts_index_text.dart
// -- Functions

String prepareFtsIndexText(String text) {
  if (text.isEmpty) {
    return text;
  }
  final output = StringBuffer();
  int? previousCodePoint;
  var previousWasCjk = false;
  for (final codePoint in text.runes) {
    final character = String.fromCharCode(codePoint);
    final isWhitespace = character.trim().isEmpty;
    final isCjk = _isCjkCodePoint(codePoint);
    if (output.length > 0 &&
        !isWhitespace &&
        previousCodePoint != null &&
        String.fromCharCode(previousCodePoint).trim().isNotEmpty &&
        (previousWasCjk || isCjk)) {
      output.write(' ');
    }
    output.write(character);
    previousCodePoint = codePoint;
    previousWasCjk = isCjk;
  }
  return output.toString().trim().replaceAll(RegExp(r'\s+'), ' ');
}

String? ftsMatchQuery(String rawQuery) {
  final prepared = prepareFtsIndexText(rawQuery.trim());
  if (prepared.isEmpty) {
    return null;
  }
  final sanitized = prepared
      .replaceAll('"', ' ')
      .replaceAll(RegExp(r'[{}()*?:^]'), ' ')
      .replaceAll(RegExp(r'\s+'), ' ')
      .trim();
  return sanitized.isEmpty ? null : '"$sanitized"';
}

bool _isCjkCodePoint(int codePoint) {
  return (codePoint >= 0x3400 && codePoint <= 0x4DBF) ||
      (codePoint >= 0x4E00 && codePoint <= 0x9FFF) ||
      (codePoint >= 0xF900 && codePoint <= 0xFAFF) ||
      (codePoint >= 0x20000 && codePoint <= 0x2FA1F) ||
      (codePoint >= 0x3040 && codePoint <= 0x30FF) ||
      (codePoint >= 0xAC00 && codePoint <= 0xD7AF);
}
